internal/controllers/lvmcluster/status.go

```diff
@@ -0,0 +1,174 @@
+	})
+}
+
+func setResourcesAvailableConditionFalse(instance *lvmv1alpha1.LVMCluster, reconcileError error) {
+	meta.SetStatusCondition(&instance.Status.Conditions, metav1.Condition{
+		Type:    lvmv1alpha1.ResourcesAvailable,
+		Status:  metav1.ConditionFalse,
+		Reason:  ReasonResourcesSyncFailed,
+		Message: fmt.Sprintf(MessageReasonResourcesSyncFailed, reconcileError),
+	})
+}
+
+func setVolumeGroupsReadyConditionTrue(instance *lvmv1alpha1.LVMCluster) {
+	meta.SetStatusCondition(&instance.Status.Conditions, metav1.Condition{
+		Type:    lvmv1alpha1.VolumeGroupsReady,
+		Status:  metav1.ConditionTrue,
+		Reason:  ReasonVGsReady,
+		Message: MessageVGsReady,
+	})
+}
+
+func setVolumeGroupsReadyConditionFailed(instance *lvmv1alpha1.LVMCluster) {
+	meta.SetStatusCondition(&instance.Status.Conditions, metav1.Condition{
+		Type:    lvmv1alpha1.VolumeGroupsReady,
+		Status:  metav1.ConditionFalse,
+		Reason:  ReasonVGsFailed,
+		Message: MessageVGsFailed,
+	})
+}
+
+func setVolumeGroupsReadyConditionDegraded(instance *lvmv1alpha1.LVMCluster) {
+	meta.SetStatusCondition(&instance.Status.Conditions, metav1.Condition{
+		Type:    lvmv1alpha1.VolumeGroupsReady,
+		Status:  metav1.ConditionFalse,
+		Reason:  ReasonVGsDegraded,
+		Message: MessageVGsDegraded,
+	})
+}
+
+func computeVolumeGroupsReadyReason(vgNodeStatusList *lvmv1alpha1.LVMVolumeGroupNodeStatusList) string {
+	reason := ReasonVGReadinessInProgress
+	vgsCount := 0
+	readyVGsCount := 0
+	for _, nodeItem := range vgNodeStatusList.Items {
+		for _, item := range nodeItem.Spec.LVMVGStatus {
+			if item.Status == lvmv1alpha1.VGStatusReady {
+				readyVGsCount++
+			} else {
+				if item.Status == lvmv1alpha1.VGStatusFailed {
+					return ReasonVGsFailed
+				} else if item.Status == lvmv1alpha1.VGStatusDegraded {
+					reason = ReasonVGsDegraded
+				}
+			}
+			vgsCount++
+		}
+	}
+
+	if readyVGsCount == vgsCount {
+		return ReasonVGsReady
+	}
+
+	return reason
+}
+
+func computeDeviceClassStatuses(vgNodeStatusList *lvmv1alpha1.LVMVolumeGroupNodeStatusList) []lvmv1alpha1.DeviceClassStatus {
+	vgNodeMap := make(map[string][]lvmv1alpha1.NodeStatus)
+	for _, nodeItem := range vgNodeStatusList.Items {
+		for _, item := range nodeItem.Spec.LVMVGStatus {
+			vgNodeMap[item.Name] = append(vgNodeMap[item.Name],
+				lvmv1alpha1.NodeStatus{
+					Node:     nodeItem.Name,
+					VGStatus: *item.DeepCopy(),
+				},
+			)
+		}
+	}
+	var allVgStatuses []lvmv1alpha1.DeviceClassStatus
+	for key, val := range vgNodeMap {
+		allVgStatuses = append(allVgStatuses,
+			lvmv1alpha1.DeviceClassStatus{
+				Name:       key,
+				NodeStatus: val,
+			},
+		)
+	}
+
+	return allVgStatuses
+}
+
+func computeLVMClusterReadiness(conditions []metav1.Condition) (lvmv1alpha1.LVMStateType, bool) {
+	state := lvmv1alpha1.LVMStatusProgressing
+	readyConditionCount := 0
+	for _, c := range conditions {
+		if c.Status == metav1.ConditionTrue {
+			readyConditionCount++
+		} else {
+			state = translateReasonToState(c.Reason, state)
```

Comment:
why do we only do this when a condition is false? shouldnt we recompute the state everytime?

Reply:
I changed it to always recompute. I also added an Unknown state as default to distinguish an unknown state from in progress. Tell me what you think.